### ISSUE TITLE
rare rewriting crash cost mismatch

### ISSUE THREAD

**Issue by mlb2251** [mlb2251/stitch#104]
`cargo run --release --bin=compress experiments/data/rec-fp/iteration=10_2019-07-11T19:49:10.899159/iteration_0.json --fmt=dreamcoder -i1 -a4 -n100`

crashes with

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `78878`,
 right: `69182`: 
[fn_0 arity=4: (#0 (#3 (#1 #2)))]
', src/extraction.rs:88:5
```

looks like a truly absurd invention that it thinks should rewrite to something better than it does.

**Comment by mlb2251** [comment]
`cargo run --release --bin=compress data/cogsci/nuts-bolts.json -i10 -a4` yields

```
Cost before: 212456
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `231646`,
 right: `193266`: 
[fn_5 arity=4: (#3 #2 #1 #0)]
', src/extraction.rs:89:9
```

**Comment by mlb2251** [comment]
seems specific to arity 4 not lower arities

**Comment by mlb2251** [comment]
we  can track it directly with
`cargo run --release --bin=compress data/cogsci/nuts-bolts.json -i1 -a4 --track="(#3 #2 #1 #0)" --follow-track --no-opt`

**Comment by mlb2251** [comment]
compressive_utility: -445006
global_correction:      590244

**Comment by mlb2251** [comment]
Omg absolutely wild. It turns out `utility_of_loc_once` was going negative meaning each usage of the invention could have negative compressive effect, then it was choosing force-reject all these usages which resulted in a massive positive correction so it could trick the system into thinking it was positive when it was super negative.

Basically it was turning the app penalty of like -104 into a positive a bunch of times haha

**Comment by mlb2251** [comment]
ah yes in particular this code was assuming that children by default would Accept and therefore that we would gain their negative utility:

```
// if it's not in corrected_utils, it must have had no conflicts so we must be switching from accept to reject with no other side effects
                // so we do (reject - accept) = (- util(idx) - 0) = - util(idx)
                // where accept was 0 since it caused no conflicts
                .unwrap_or_else(|| - utility_of_loc_once[*idx]) 
```

**Comment by mlb2251** [comment]
fixed in ebd7199ab73464fdf51958b00432de078d1aa121
